byron/ledger/executable-spec/src/Ledger/Delegation.hs

```diff
@@ -339,30 +361,53 @@ instance STS ADELEG where
     | NoLastDelegation
       -- | Not a failure; this should just pass the other rule
     | AfterExistingDelegation
+    -- | The given key is a delegate of the given genesis key.
+    | AlreadyADelegateOf VKey VKeyGenesis
+
     deriving (Eq, Show)
 
-  initialRules = [ do
-                     IRC env <- judgmentContext
-                     return DState
-                       { _dStateDelegationMap  = Bimap.fromList $ map (\vkg@(VKeyGenesis key) -> (vkg, key)) (Set.toList env)
-                       , _dStateLastDelegation = Map.fromSet (const (Slot 0)) env
-                       }
-                 ]
+  initialRules = [
+    do
+      IRC env <- judgmentContext
+      return DState
+        { _dStateDelegationMap  =
+            Bimap.fromList
+            $ map (\vkg@(VKeyGenesis key) -> (vkg, key)) (Set.toList env)
+        , _dStateLastDelegation = Map.fromSet (const (Slot 0)) env
+        }
+    ]
   transitionRules =
     [ do
-        TRC (_env, st, (slt, (vks, vkd))) <- judgmentContext
-        case Map.lookup vks (st ^. lastDelegation) of
-          Nothing -> True ?! BeforeExistingDelegation
-          Just sp -> sp < slt ?! BeforeExistingDelegation
-        return $ st
-          & delegationMap %~ (\sdm -> sdm ⨃ [(vks, vkd)])
-          & lastDelegation %~ (\ldm -> ldm ⨃ [(vks, slt)])
+        TRC ( _env
+            , DState { _dStateDelegationMap = dms
+                     , _dStateLastDelegation = dws
+                     }
+            , (s, (vks, vkd))
+            ) <- judgmentContext
+        vkd ∉ range dms ?! AlreadyADelegateOf vkd (dms !> vkd)
+        case Map.lookup vks dws of
+          Nothing -> pure () -- If vks hasn't delegated, then we proceed and
+                             -- update the @ADELEG@ state.
+          Just sp -> sp < s ?! BeforeExistingDelegation
+        return $!
+          DState { _dStateDelegationMap = dms ⨃ [(vks, vkd)]
+                 , _dStateLastDelegation = dws ⨃ [(vks, s)]
+                 }
```

Comment:
You trying to get rid of lenses?

Reply:
Oh, no, you just needed the value somewhere else

Reply:
> You trying to get rid of lenses?

 Shhhh... :stuck_out_tongue: 

Yeah, that's what I would like to discuss. I don't think lenses make sense for the executable spec, and I like the style in `cardano-ledger` :)